### DISCUSSION TITLE
Caller, how to get the correct line number

### DISCUSSION THREAD

**Discussion by LazarenkoA** [uber-go/zap#1154] [General]
hi, I have such a project structure
..
|_main.go
|_logger
|___logger.go

body logger.go:
<details> 

```go
package logger

import (
	"go.uber.org/zap"
	"go.uber.org/zap/zapcore"
)

var sugar *zap.SugaredLogger

func SetLogger() {
	cfg := zap.Config{
		Level:             zap.NewAtomicLevelAt(zapcore.InfoLevel),
		Development:       false,
		DisableCaller:     true,
		DisableStacktrace: true,
		Encoding:          "json",
		EncoderConfig: zapcore.EncoderConfig{
			MessageKey:    "msg",
			LevelKey:      "level",
			NameKey:       "name",
			EncodeLevel:   zapcore.CapitalLevelEncoder,
			TimeKey:       "time",
			EncodeTime:    zapcore.RFC3339TimeEncoder,
			CallerKey:     "caller",
			EncodeCaller:  zapcore.ShortCallerEncoder,
			StacktraceKey: "stacktrace",
			EncodeName:    zapcore.FullNameEncoder,
			LineEnding:    "\n",
		},
		OutputPaths:      []string{"stdout"},
		ErrorOutputPaths: []string{"stderr"},
	}

	logger, _ := cfg.Build()
	logger = logger.WithOptions(
		zap.AddCaller(),
		zap.AddStacktrace(zap.ErrorLevel),
	)
	sugar = logger.Sugar()
}

func Info(msg string) {
	sugar.Info(msg)
}

```
</details>

main.go:
```go

package main

import (
	"test/logger"
)

func init() {
	logger.SetLogger()
}

func main() {
	test()
}

func test() {
	test2()
}

func test2() {
	logger.Info("test")
}


```

result:
`{"level":"INFO","time":"2022-08-19T22:22:42+03:00","caller":"logger/logger.go:44","msg":"test"}`
`"caller":"logger/logger.go:44"` - not correct

to fix this, you need to add option `zap.AddCallerSkip(1)` after that, the result is 
`{"level":"INFO","time":"2022-08-19T22:25:41+03:00","caller":"test/main.go:21","msg":"test"}`
ok, but what if in program use local logger and  global logger 

```go
func test2() {
	l := logger.Sugar.With("value", 2)
	l.Info("test")
}
```
it turns out that the caller is incorrect again
`{"level":"INFO","time":"2022-08-19T22:33:26+03:00","caller":"test/main.go:16","msg":"test","value":2}`
what can you come up with in this situation so that the caller is always correct?

**Comment by LazarenkoA** [comment]
I wrote a question and came up with a solution, in principle, this is no longer an actual question

**Comment by CaioTeixeira95** [comment]
Hello 👋 

Couldn't you only do this?

```go
func test2() {
    l := logger.WithOptions(zap.AddCallerSkip(-1)).Sugar()
    l.Info("test")
}
```

**Comment by LazarenkoA** [comment]
thanks
